pkg/advisory/advisory.go

```diff
@@ -140,89 +128,62 @@ func Latest(entries []build.AdvisoryContent) *build.AdvisoryContent {
 	return &latestEntry
 }
 
-func NewConfigUpdaterForAdvisories(i *configs.Index, transform func(build.Advisories) (build.Advisories, error)) configs.UpdateFunc {
-	return i.NewUpdater(func(node *yaml.Node) error {
-		advNode, err := yamlNodeForAdvisories(node)
-		if err != nil {
-			return err
-		}
-
-		advisories := build.Advisories{}
-		err = advNode.Decode(&advisories)
-		if err != nil {
-			return err
-		}
-
-		updatedAdvisories, err := transform(advisories)
-		if err != nil {
-			return err
-		}
-
-		err = advNode.Encode(updatedAdvisories)
-		if err != nil {
-			return err
-		}
+type DiscoverOptions struct {
+	// The Index of package configs on which to operate.
+	Index *configs.Index
 
-		return nil
-	})
+	// VulnerabilitySearcher is how Discover searches for vulnerabilities to match to
+	// packages.
+	VulnerabilitySearcher vuln.Searcher
 }
 
-func NewConfigUpdaterForSecfixes(i *configs.Index, transform func(secfixes build.Secfixes) (build.Secfixes, error)) configs.UpdateFunc {
-	return i.NewUpdater(func(node *yaml.Node) error {
-		sfNode, err := yamlNodeForSecfixes(node)
-		if err != nil {
-			return err
-		}
-
-		secfixes := build.Secfixes{}
-		err = sfNode.Decode(&secfixes)
-		if err != nil {
-			return err
-		}
+// Discover searches for new vulnerabilities that match packages in a config
+// index, and adds new advisories to configs for vulnerabilities that haven't
+// been noted yet.
+func Discover(options DiscoverOptions) error {
+	matches, err := options.VulnerabilitySearcher.AllVulnerabilities()
+	if err != nil {
+		return fmt.Errorf("unable to discover advisories: %w", err)
+	}
 
-		updatedSecfixes, err := transform(secfixes)
-		if err != nil {
-			return err
+	err = options.Index.Select().UpdateAdvisories(func(cfg build.Configuration) (build.Advisories, error) {
+		matchesForPackage := matches[cfg.Package.Name]
+		if len(matchesForPackage) == 0 {
+			// nothing to update for this package
+			return build.Advisories{}, configs.ErrSkip
 		}
 
-		err = sfNode.Encode(updatedSecfixes)
-		if err != nil {
-			return err
+		//nolint:gocritic // rangeValCopy rule not worth it here
+		for _, m := range matchesForPackage {
+			if !m.CPE.VersionRange.Includes(cfg.Package.Version) {
+				continue
+			}
+
+			vulnID := m.Vulnerability.ID
+			_, exists := cfg.Advisories[vulnID]
+
+			if exists {
+				// TODO: Should we allow for updating existing advisories if previously read vuln
+				// data has been updated?
```

Comment:
cant help but think a log line here will help later, but maybe its too verbose

Reply:
Nice — added in becec34242f0616adea99e55ab1cac0b060225c4!